content/articles/working-with-java-synchronization/index.md

```diff
@@ -0,0 +1,294 @@
+
+For example, if checks against a checking account are processed consecutively, the system will first check for sufficient money to process check A, then check again for sufficient funds to process check B. If the two checks are processed simultaneously, the system may interpret the same account balance for both transactions producing an overdraft.
+
+2. **Check-then-act condition:** This race scenario occurs when two threads/processes verify the same value for an external operation. Both threads/processes check the value, but only one can accept it. The subsequent thread/process will read it as null. As a result, the program's next action is determined by an out-of-date or unavailable observation. For example, map applications that need the same location data can't use each other's data if they're running at the same time. The data is treated as null in the subsequent stage.
+
+The example program below illustrates race conditions:
+
+You can see the increased value of an integer variable in this basic example. One by one the variable's value is incremented and shown in ten different ways. Each thread will be numbered from one to nine.
+
+```java
+public class Example {
+  int check = 0;
+  public  void CheckBoost(){
+    try {
+      Thread.sleep(1000);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+    check++;
+  }
+  public int obtainCheck(){
+    return check;
+  }
+  public static void main(String[] args) {
+    Example zy = new Example();
+    for(int x = 1; x < 6; x++){
+      new Thread(new Runnable() {			
+        @Override
+        public void run() {
+          zy.CheckBoost();
+          System.out.println("The output of the thread is : " + Thread.currentThread().getName() + " - " + zy.obtainCheck());
+        }
+      }).start();
+    }
+  }
+}
+```
+
+```
+The output of the thread is : Thread-0 - 3
+The output of the thread is : Thread-1 - 3
+The output of the thread is : Thread-2 - 3
+The output of the thread is : Thread-4 - 5
+The output of the thread is : Thread-3 - 4
+```
+
+As shown above, the threads are chosen at random, and the value is incorrect. The value should rise by 1 but that is not the case. Usually, the output is 3. Threads 0, 1, and 2 share the same value thus showing a race condition. After understanding what a race condition is, let's now look at how to refrain from it.
+
+It is evident that the crucial element (code that changes shared resources) must be limited. Additionally, with Java's `synchronized` keyword we can synchronize access to the shared resource. This prevents thread interference during atomic operations. The term "atomic operation" refers to a set of operations that are always performed in unison. All of the atomic actions must be completed at the same time, or none of them can be completed at all.

```

Comment:
I meant atomic operations cannot be done when there is interference since they all should complete before other processes start. So the synchronized keyword synchronizes the access to the shared resource.

Reply:
I'm not sure what you mean by this. How does the synchronized keyword and how it synchronizes access to the shared resource relate to atomic operations?